src/Tasks/Microsoft.NET.Build.Tasks/DiagnosticsHelper.cs

```diff
@@ -0,0 +1,153 @@
+                message,
+                filePath,
+                severity,
+                startLine,
+                startColumn,
+                endLine: startLine,
+                endColumn: startColumn,
+                targetFrameworkMoniker: targetFrameworkMoniker,
+                packageId: packageId,
+                logToMSBuild: logToMSBuild);
+
+        public ITaskItem Add(
+            string diagnosticCode,
+            string message,
+            string filePath,
+            DiagnosticMessageSeverity severity,
+            int startLine,
+            int startColumn,
+            int endLine,
+            int endColumn,
+            string targetFrameworkMoniker,
+            string packageId,
+            bool logToMSBuild = true)
+        {
+            string itemspec =
+                (string.IsNullOrEmpty(targetFrameworkMoniker) ? string.Empty : $"{targetFrameworkMoniker}/") +
+                (string.IsNullOrEmpty(packageId) ? string.Empty : $"{packageId}/") + 
+                diagnosticCode;
+
+            var diagnostic = new TaskItem(itemspec, new Dictionary<string, string>
+            {
+                { MetadataKeys.DiagnosticCode, diagnosticCode },
+                { MetadataKeys.Message,        message },
+                { MetadataKeys.FilePath,       filePath ?? string.Empty },
+                { MetadataKeys.Severity,       severity.ToString() },
+
+                { MetadataKeys.StartLine,      startLine.ToString() },
+                { MetadataKeys.StartColumn,    startColumn.ToString() },
+                { MetadataKeys.EndLine,        endLine.ToString() },
+                { MetadataKeys.EndColumn,      endColumn.ToString() },
+
+                { MetadataKeys.ParentTarget,   targetFrameworkMoniker ?? string.Empty },
+                { MetadataKeys.ParentPackage,  packageId ?? string.Empty },
+            });
+
+            _diagnosticMessages.Add(diagnostic);
+
+            if (logToMSBuild)
+            {
+                LogToMSBuild(diagnosticCode, message, filePath, severity, startLine, startColumn, endLine, endColumn);
+            }
+
+            return diagnostic;
+        }
+
+        private void LogToMSBuild(string diagnosticCode, string message, string filePath, DiagnosticMessageSeverity severity, int startLine, int startColumn, int endLine, int endColumn)
+        {
+            switch (severity)
+            {
+                case DiagnosticMessageSeverity.Error:
+                    _log.LogError(
+                        subcategory     : null,
+                        errorCode       : diagnosticCode,
+                        helpKeyword     : null,
+                        file            : filePath, 
+                        lineNumber      : startLine,
+                        columnNumber    : startColumn,
+                        endLineNumber   : endLine,
+                        endColumnNumber : endColumn, 
+                        message         : message);
```

Comment:
Nit: Does VS preserve this formatting on Ctrl+K, Ctrl + F? 

I'd prefer we did not get cute lining things up against the grain. It causes unnecessary diffs when something longer than all of the others gets added.